🔧 MAINTAIN: Add mypy type-checking

hukkinj1 opened the pull request:
Closes https://github.com/executablebooks/markdown-it-py/issues/19

hukkinj1 commented on markdown_it/common/utils.py:
```diff
@@ -141,9 +141,10 @@ def replaceEntityPattern(match, name):
+    raise NotImplementedError
+    # if "\\" in string:
+    #     return string
+    # return string.replace(UNESCAPE_MD_RE, "$1")
```
I simply commented this function out, because `str.replace()` takes strings as input, not re.Patterns, so this is probably broken.

chrisjsewell commented:
I will get round to looking at this eventually! Just want to make sure I look thoroughly, since its touching core code

hukkinj1 commented:
I rebased on master, and added one more commit that adds `no_implicit_optional = True` mypy configuration. This is to conform to [updated PEP 484](https://github.com/python/peps/pull/689). Eventually this will be default mypy configuration too.

chrisjsewell commented:
👍 

> this will be default mypy configuration too.

TBH it would be so much nicer and concise if you could use the typescript `?` notation:

```python
def func(a: Employee? = None):
  pass
```

hukkinj1 commented:
I think I've seen some discussion around that `Optional` syntax somewhere. Python typing is moving forward at quite a rapid pace so maybe one day we'll see that. In the next Python release we already get [union types with `TypeX | TypeY`](https://www.python.org/dev/peps/pep-0604/) which I really like :)

chrisjsewell commented:
Oh thats nice 😄 Yeh for typing, I think they should just copy everything from typescript, which they are basically gradually doing

hukkinj1 commented:
Haha yeh seems so. Btw. the new union syntax obviously lets us do `int | None` which is more concise than `Optional[int]` and requires no import from `typing`, so partially solves the problem.
